### PR TITLE
Improve multipartUpload & S3 Test support metadata & content-type

### PR DESCRIPTION
**Pull request by regis-leray**
- S3 Test - support metadata & content type
-  MultipartUpload validate parrallelism & Partsize

### CODE REVIEW

**Comment by regis-leray** [comment]
@melgenek I made a few changes. 

Parallelism is now a mandatory field since this has a big impact on behavior and performance. The default value is useless since it depends on the specific use case. I let the user decide what is best.

I defined PartSize as a type since I prefer to rely on compilation than an error runtime error.

And I also support metadata in the Test module.

The next PR is to use ZIO test generator (aka scala-check), and remove all manually generated values.

File: src/main/scala/zio/s3/Live.scala [hunk 1]
```diff
@@ -130,9 +130,17 @@ final class Live(unsafeClient: S3AsyncClient) extends S3.Service {
     bucketName: String,
     key: String,
     content: ZStream[R, Throwable, Byte],
-    options: MultipartUploadOptions = MultipartUploadOptions()
-  ): ZIO[R, S3Exception, Unit] =
+    options: MultipartUploadOptions
+  )(parallelism: Int): ZIO[R, S3Exception, Unit] =
     for {
+      _        <- ZIO.cond(
```
Comment (melgenek) [inline src/main/scala/zio/s3/Live.scala:136]: I would have the same validation for the part size without any additional types.

File: src/test/scala/zio/s3/S3Test.scala [hunk 2]
```diff
@@ -198,24 +197,71 @@ object S3Suite {
...
+          objectMetadata <- getObjectMetadata(bucketName, tmpKey) <* ZFiles.delete(root / bucketName / tmpKey)
+        } yield assert(objectMetadata.contentType)(equalTo("application/json")) &&
+          assert(objectMetadata.metadata.map { case (k, v) => k.toLowerCase -> v })(equalTo(Map("key1" -> "value1")))
+      },
+      testM("multipart object when the chunk size and parallelism are customized") {
+        val (dataSize, data) = randomNEStream
+        val tmpKey           = Random.alphanumeric.take(10).mkString
+
+        for {
+          _partSize     <- PartSize.from(10 * PartSize.Mega)
```
Comment (melgenek) [inline src/test/scala/zio/s3/S3Test.scala:259]:
```suggestion
          partSize     <- PartSize.from(10 * PartSize.Mega)
```

**Review by melgenek — COMMENTED** [review]
@regis-leray 
1) Regarding the part size type, I tried to explain why I am against having an additional type here https://github.com/zio/zio-s3/pull/117#discussion_r487441005. In short: the more configurable fields there are in the upload method, the more changes the user of the library has to do in their own codebase. As a user, I would prefer the library doing this instead of me.

Regarding the PartSize and validation, I would suggest summoning somebody from the zio-sqs maintainers to have a somewhat consistent API. In zio-sqs there are configurations, like the number of messages that have to be received from the sqs. The range is 1-10 but it's still represented as a plain Int https://github.com/zio/zio-sqs/blob/7b6a2e44985d43fbc0090d24967414a84b3a0d65/src/main/scala/zio/sqs/SqsStreamSettings.scala#L5

My preference: keep the part size config as Int.

2) I would put the sane defaults to all the non-required fields. Having more fields to configure is not always better.
I would have the parallelism=1 by default, while the code still even if it is not so efficient as with more parallelism.

3) Regarding moving the parallelism out of the MultipartUploadOptions to the method signature: I would keep it in the MultipartUploadOptions. I just don't see a difference compared, for example, to the part size. They are both relevant for the execution. Why should these configuration params be passed differently?

**Comment by regis-leray** [comment]
1. Creating PartSize, is to makee obvious to the user the operation could faild if the provided value is wrong.

in the current implementation the option api let the user believe he create a valid options, which is not the case !
It will fail at runtime
```
val options = MultipartUploadOptions(partSize = 1024)
```

2. Parallelism is now a mandatory field since this has a big impact on behavior and performance. Since we cannot pick a valid default value, i prefer to let the user decided. As example `mapParN` doesnt have a default value. what we can do is to provide an alias ;)

```scala
def multipartUpload_[R <: Has[_]: Tag](
                                         bucketName: String,
                                         key: String,
                                         content: ZStream[R, Throwable, Byte],
                                         options: MultipartUploadOptions = MultipartUploadOptions()
                                       )
```

**Comment by regis-leray** [comment]
@melgenek I just added a way to modify PartSize in a safe way :)
```
MultipartOptions(partSize = PartSize.Min * 2)  //10Mb
MultipartOptions(partSize = PartSize.Min + (200 * Kilo))  //5.2kb
```

File: src/main/scala/zio/s3/Live.scala (hunk 1 above)
Comment (regis-leray) [reply src/main/scala/zio/s3/Live.scala:136]: i follow your suggestion

**Comment by melgenek** [comment]
@regis-leray Thank you for the improvements!

**Comment by regis-leray** [comment]
@melgenek 
https://github.com/zio/zio-s3/releases/tag/v0.2.4

**Pull request merged**